# nf-core/hicar: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

* [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipeline
* [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution
* [genome](#genome) - The genome related files used in the pipeline.
* [checksums](#checksums) - The checksums (md5) of decompressed fastq files.
* [FastQC](#fastqc) - Raw read QC
* [bwa](#alignment) - The alignments of trimmed reads.
* [pairs](#pairs) - The interaction pairs and their quality control reports.
* [cooler](#cooler) - Cooler files for visualization
* [ATACpeak](#call-peaks-for-short-reads) - The peaks called ATAC reads (R2 reads).
* [MAPSpeak](#tads-and-loops) - The TADs and loops determined by MAPS.
* [DifferentialAnalyis](#differential-analysis) - Differential analysis for the TADs and loops.
* [igv.js](#igv) - The track files which can be viewed by web-server.

### MultiQC

<details markdown="1">
<summary>Output files</summary>

* `multiqc/`
    * `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
    * `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
    * `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

* `pipeline_info/`
    * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
    * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.tsv`.
    * Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

### checksums

<details markdown="1">
<summary>Output files</summary>

* `checksums/*.txt`: md5 checksums.

</details>

The checksums (md5) of decompressed fastq files. Those infomation can be used for GEO submission.

### Genome

<details markdown="1">
<summary>Output files</summary>

* `genome/`
    * `genome.fa.sizes`: chromosome sizes file
    * `genome.fa.fai`: genome index file
    * `filtered/genome.include_regions.bed`: filtered genome by blacklist
    * `digest/*`: genomic features files digested by given restriction enzyme

</details>

If the mappability file is not provided in the profile file, the [GenMap](https://pubmed.ncbi.nlm.nih.gov/32246826/) will be used to create the mappability file with read length set to 50.

### FastQC

<details markdown="1">
<summary>Output files</summary>

* `fastqc/`
    * `*_fastqc.html`: FastQC report containing quality metrics.
    * `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

> **NB:** The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.

### Alignment

<details markdown="1">
<summary>Output files</summary>

* `bwa/mapped/bam/*.(bam|bai)`: The files resulting from the alignment of individual libraries will contain the coordinate sorted alignment files in [`*.bam`](https://samtools.github.io/hts-specs/SAMv1.pdf) format.
* `bwa/mapped/QC/*`: The stats of mapping results.

</details>

Adapter-trimmed reads are mapped to the reference assembly using [BWA::mem](http://bio-bwa.sourceforge.net/bwa.shtml). A genome index is required to run BWA so if this is not provided explicitly using the `--bwa_index` parameter then it will be created automatically from the genome fasta input. The index creation process can take a while for larger genomes.

### Pairs

<details markdown="1">
<summary>Output files</summary>

* `pairs/`
    * `raw/*`: The raw reads pairs for each sample and stats for the pairs.
    * `filtered/*`: The filtered files. The unselected.pairs.gz files in samefrag folder can be used to generate virtual 4C plots.
    * `QC/*`: The quality analysis results.

</details>

The raw and filtered pairs exported by [pairtools](https://pairtools.readthedocs.io/en/latest/).
The quality analysis for filtered pairs were done by [pairsqc](https://github.com/4dn-dcic/pairsqc).

### Cooler

<details markdown="1">
<summary>Output files</summary>

* `cooler/mcool/*`: The mcool files for each group.
* `cooler/hic/*`: The .hic files for each group.

</details>

[Cooler](https://cooler.readthedocs.io/en/latest/index.html) files for visualization.
To view the `.mcool` files, you may want to try [higlass-manage](https://github.com/higlass/higlass-manage).
Once the `higlass-manage` is install, view the files by

```
higlass-manage view cooler/mcool/your.mcool
```

[Juicebox](https://github.com/aidenlab/Juicebox) can be used to view the `.hic` file.
`.hic` file is the second choices for visualization in case installation of `higlass-manage` is impossible.

### Call peaks for short reads

<details markdown="1">
<summary>Output files</summary>

* `ATACpeak/`
    * `R2_bigwig/*`: The bigWig files of R2_reads.
    * `peaks_per_Group/*`: The called peaks for each group.
    * `merged_peaks/*`: The merged peaks for all groups.

</details>

The peaks called ATAC reads (R2 reads) by [MACS2](https://github.com/macs3-project/MACS).

### TADs and loops

<details markdown="1">
<summary>Output files</summary>

* `MAPSpeak/`
    * `bin*/*`: The annotated TADs and loops.
    * `*`: The tables of TADs and loops with p-value and fdr.

</details>

The TADs and loops are called by [MAPS](https://pubmed.ncbi.nlm.nih.gov/30986246/).
MAPS is a model-based analysis of long-range chromatin interactions.

### Differential analysis

<details markdown="1">
<summary>Output files</summary>

* `DifferentialAnalyis/*`: Differential analysis results.

</details>

Differential analysis for the TADs and loops by [edgeR](https://pubmed.ncbi.nlm.nih.gov/19910308/).
The input counts are filtered counts dumped by pairtools for each samples.
The differential analysis is done for validated TADs and loops by MAPs.
Annotation is done by [ChIPpeakAnno](https://pubmed.ncbi.nlm.nih.gov/20459804/)
for the overlap features (gene level) or nearest features.

### IGV

<details markdown="1">
<summary>Output files</summary>

* `igv.js/*`: index.html, readme.txt and track.tgz.

</details>

The track.tga contain the R2 reads tracks and interactions called by MAPS. Untar the `track.tgz` file and copy all the files into same folder in a web-server to view the data.
